cpp/src/arrow/dataset/dataset_encryption_test.cc

```diff
@@ -0,0 +1,391 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <string_view>
+#include "arrow/api.h"
+#include "arrow/array/builder_primitive.h"
+#include "arrow/builder.h"
+#include "arrow/dataset/api.h"
+#include "arrow/dataset/parquet_encryption_config.h"
+#include "arrow/dataset/partition.h"
+#include "arrow/filesystem/mockfs.h"
+#include "arrow/io/api.h"
+#include "arrow/status.h"
+#include "arrow/table.h"
+#include "arrow/testing/gtest_util.h"
+#include "gtest/gtest.h"
+#include "parquet/arrow/reader.h"
+#include "parquet/encryption/test_in_memory_kms.h"
+
+constexpr std::string_view dsFooterMasterKey = "0123456789012345";
+constexpr std::string_view dsFooterMasterKeyId = "footer_key";
+constexpr std::string_view dsColumnMasterKeys[] = {"1234567890123450"};
+constexpr std::string_view dsColumnMasterKeyIds[] = {"col_key"};
+const int dsNumColumns = 1;
+namespace arrow {
+namespace dataset {
+
+class DatasetEncryptionTest : public ::testing::Test {
+ protected:
+  // setup the test
+  void SetUp() {}
+
+  // Create our parquetfileformat with encryption properties
+  std::shared_ptr<ParquetFileFormat> CreateFileFormat(
+      const std::string_view* column_ids, const std::string_view* column_keys,
+      int num_columns, const std::string_view& footer_id,
+      const std::string_view& footer_key,
+      const std::string_view& footer_key_name = "footer_key",
+      const std::string_view& column_key_mapping = "col_key: a") {
+    auto key_list =
+        BuildKeyMap(column_ids, column_keys, num_columns, footer_id, footer_key);
+
+    auto crypto_factory = SetupCryptoFactory(true, key_list);
+    ::parquet::encryption::KmsConnectionConfig kms_connection_config;
+    auto encryption_config =
+        std::make_shared<::parquet::encryption::EncryptionConfiguration>(
+            std::string(footer_key_name));
+    encryption_config->column_keys = column_key_mapping;
+    if (footer_key_name.size() > 0) {
+      encryption_config->footer_key = footer_key_name;
+    }
+
+    DatasetEncryptionConfiguration dataset_encryption_config{
+        crypto_factory,
+        std::make_shared<parquet::encryption::KmsConnectionConfig>(kms_connection_config),
```

Comment:
It seems `kms_connection_config` and `decryption_config` are using default values. It may be much easier to use if user does not have to pass a default value.